contracts/cw721-base/src/lib.rs

```diff
@@ -4,3 +4,43 @@ pub mod msg;
 pub mod state;
 
 pub use crate::error::ContractError;
+
+pub mod entry {
+    use super::*;
+
+    #[cfg(not(feature = "library"))]
+    use cosmwasm_std::entry_point;
+    use cosmwasm_std::{Binary, Deps, DepsMut, Empty, Env, MessageInfo, Response, StdResult};
+
+    // This is a simple type to let us handle empty extensions
+    pub type Extension = Option<Empty>;
+
+    // This makes a conscious choice on the various generics used by the contract
+    #[cfg_attr(not(feature = "library"), entry_point)]
+    pub fn instantiate(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: msg::InstantiateMsg,
+    ) -> StdResult<Response> {
+        let tract = contract::Cw721Contract::<Extension>::default();
+        tract.instantiate(deps, env, info, msg)
+    }
+
+    #[cfg_attr(not(feature = "library"), entry_point)]
+    pub fn execute(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: msg::ExecuteMsg<Extension>,
+    ) -> Result<Response, ContractError> {
```

Comment:
I was thinking about the problem of handling multiple types of messages (like base message and extension message), and I got a revelation - this is a place. As it is just a boilerplate it makes sense to instead of take `msg` as `msg::ExecuteMsg<Extension>`, take it here as something generic, before it is parsed (like `Binary`), and then just make deserialization here, like (this is more pseudocode):
```rust
let contract = contract::Cw721Contract::default();
if let Ok(msg) = deserialize::<msg::ExecuteMsg<Extension>>() {
  Cw721::execute(&contract, deps, env, info, msg)
} else if let Ok(msg) = deserialize::<extension::ExecuteMsg>() {
  Extension::execute(&contract, deps, env, info, msg)
} else {
  return Err(StdError::InvalidMessage.into());
}
```

The only problem I have here is if there is any type which is cheap/free to deserialize into by `entry_point` (so deserialization is not performed twice). In future, when this is generated it can even generate raw entry point.

Reply:
I would auto-generate some ser/deserialization code that does that.

eg.

```rust
enum ExtendedExecMsg {
  Cw721(cw721::ExecMsg),
  My(crate::ExecMsg),
}
```

And auto-generate ser/deser such that it flattens them, basically doing what you do above - try to deter into cw721::ExecMsg, if that fails, deser into My.

This is a subset of the serde "flatten" option, and maybe you want to make a macro to implement Serialize on this? Or maybe we try it first with a manually implemented Serialize/Deserialize and see if it makes sense.